### PR TITLE
disabled duplicates in the queue

### LINKED ISSUE #20: Duplicate videos can be added to the queue
I can't think of a reasonable use case where a user would want to add duplicate videos to the queue.

### PR DESCRIPTION
**Pull request by elenirotsides**
Disabled the ability to add duplicate videos to the queue. I did this by comparing the video ids in the created hook, and then disabled the button if the video was a duplicate so the addToQueue function couldn't be triggered. Closes #20﻿

### CODE REVIEW

File: src/components/VideoQueueItem.vue [hunk 1]
```diff
@@ -50,8 +50,23 @@ export default {
...
 		};
-	},
+	}, 
+	created() { 
+        if (this.item.id === this.$store.state.room.currentSource.id) {
+			this.isLoadingAdd = false;
+			this.hasBeenAdded = true;
+			this.existsInQueue = true;
+		}
+		for (let i = 0; i < this.$store.state.room.queue.length; i++) {
```
Comment (dyc3) [inline src/components/VideoQueueItem.vue:62]: Use a `for...of` loop here instead

File: src/components/VideoQueueItem.vue [hunk 2]
```diff
@@ -50,8 +50,23 @@ export default {
...
 			thumbnailHasError: false,
+			existsInQueue: false,
 		};
-	},
+	}, 
+	created() { 
+        if (this.item.id === this.$store.state.room.currentSource.id) {
+			this.isLoadingAdd = false;
+			this.hasBeenAdded = true;
+			this.existsInQueue = true;
```
Comment (dyc3) [inline src/components/VideoQueueItem.vue:60]: should return early

File: src/components/VideoQueueItem.vue [hunk 3]
```diff
@@ -50,8 +50,23 @@ export default {
 			isLoadingVote: false,
 			hasBeenAdded: false,
 			thumbnailHasError: false,
+			existsInQueue: false,
 		};
-	},
+	}, 
+	created() { 
+        if (this.item.id === this.$store.state.room.currentSource.id) {
+			this.isLoadingAdd = false;
```
Comment (dyc3) [inline src/components/VideoQueueItem.vue:58]: not needed

File: src/components/VideoQueueItem.vue [hunk 4]
```diff
@@ -50,8 +50,23 @@ export default {
...
+			this.hasBeenAdded = true;
+			this.existsInQueue = true;
+		}
+		for (let i = 0; i < this.$store.state.room.queue.length; i++) {
+			if (this.item.id === this.$store.state.room.queue[i].id) {
+				this.isLoadingAdd = false;
```
Comment (dyc3) [inline src/components/VideoQueueItem.vue:64]: not needed

**Review by dyc3 — CHANGES_REQUESTED** [review]
`existsInQueue` is functionally equivalent to `hasBeenAdded`, so use that instead. The video's `service` is not being checked either. A video on 2 different services _could_ have the same id. (It's an extremely rare edge case, but one that is easily avoided)

(Also, make sure you don't have trailing or inconsistent whitespace. I fixed it for you this time tho)

File: src/components/VideoQueueItem.vue (hunk 1 above)
Comment (dyc3) [reply src/components/VideoQueueItem.vue:62]: The loop should also return early

File: src/components/VideoQueueItem.vue (hunk 2 above)
Comment (elenirotsides) [reply src/components/VideoQueueItem.vue:60]: What do you mean should return early?

File: src/components/VideoQueueItem.vue (hunk 1 above)
Comment (elenirotsides) [reply src/components/VideoQueueItem.vue]: I added a break, lmk if I didn't do this correctly though

**Comment by elenirotsides** [comment]
@dyc3 Can you take a look at my latest commit when you get a chance, if you don't mind? I think I'm on the right track, but I don't know how to finish off my changes in `api.js`. Also, I'm not 100% confident that what I did is in the correct location either.

File: src/components/VideoQueueItem.vue (hunk 2 above)
Comment (dyc3) [reply src/components/VideoQueueItem.vue]: It should `return` here, because the for loop doesn't need to run

File: roommanager.js [hunk 5]
```diff
@@ -225,6 +225,10 @@ class Room {
 				this.log.error(`Failed to get video info: ${err}`);
 				queueItem.title = queueItem.id;
 			}).then(() => {
+				if (this.queue.id.includes(queueItem.id) && this.queue.service.includes(queueItem.service)) {
```
Comment (dyc3) [inline roommanager.js:228]: The condition should look more like: (https://lodash.com/docs/4.17.15#find)
```suggestion
				if (_.find(this.queue, queueItem)) {
```

And this check should go _before_ `getVideoInfo`

File: api.js [hunk 6]
```diff
@@ -43,6 +43,19 @@ function handleGetRoomFailure(res, err) {
 	}
 }
 
+function handlePostVideoFailure(res, err) {
+	/*
+	not sure what to do with this function, I know that I'm supposed to call this function in 
+	the router.post("/room/:name/queue" ..... route but I don't know where to do it
+	*/
```
Comment (dyc3) [inline api.js:50]: add `.catch(err => handlePostVideoFailure(res, err));` after the `then` calls

File: roommanager.js [hunk 7]
```diff
@@ -743,6 +746,13 @@ class RoomNameTakenException extends Error {
 	}
 }
 
+class VideoAlreadyQueuedException extends Error {
+	constructor(video) {
+		super(`The video "${video}" is already in the queue`);
```
Comment (dyc3) [inline roommanager.js:751]:
```suggestion
	constructor(title) {
		super(`The video "${title}" is already in the queue`);
```

**Review by dyc3 — APPROVED** [review]

**Pull request merged**